ios/core/Tests/HeadlessPlayerTests.swift

```diff
@@ -272,6 +272,101 @@ class HeadlessPlayerTests: XCTestCase {
         player.start(flow: FlowData.COUNTER) { _ in}
         wait(for: [updateExp], timeout: 1)
     }
+
+    func testConstantsController() {
+        let player = HeadlessPlayerImpl(plugins: [])
+        
+        guard let constantsController = player.constantsController else { return }
+        
+        // Basic get/set tests
+        let data: Any = [
+            "firstname": "john",
+            "lastname": "doe",
+            "favorite": [
+                "color": "red"
+            ]
+        ]
+        
+        constantsController.addConstants(data: data, namespace: "constants")
+
+        let firstname: String? = constantsController.getConstants(key: "firstname", namespace: "constants")
+        XCTAssertEqual(firstname, "john")
+
+        let middleName: String? = constantsController.getConstants(key:"middlename", namespace: "constants")
+        XCTAssertEqual(middleName, "undefined")
+
+        let middleNameSafe: String? = constantsController.getConstants(key:"middlename", namespace: "constants", fallback: "A")
+        XCTAssertEqual(middleNameSafe, "A")
+
+        let favoriteColor: String? = constantsController.getConstants(key:"favorite.color", namespace: "constants")
+        XCTAssertEqual(favoriteColor, "red")
+
+        let nonExistantNamespace: String? = constantsController.getConstants(key:"test", namespace: "foo")
+        XCTAssertEqual(nonExistantNamespace, "undefined")
```

Comment:
we probably shouldn't be able to get `undefined` back as a string, in Swift we should coalesce this to `nil`

`JSValue` has a property to indicate whether or not it is `undefined` -> https://developer.apple.com/documentation/javascriptcore/jsvalue/isundefined